### PR TITLE
crimson: bring the core of op tracking

### PR DESCRIPTION
**Pull request by rzarzynski**
This is a reincarnation of #39094.

TODO:
  * [ ] external logging backends like LTTng (the interfaces are already laid down),
  * [ ] dumping in-flight ops (aka serialization of the `InternalBackend`,
  * [ ] historic / slow ops backend.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/common/operation.h [hunk 1]
```diff
@@ -155,13 +158,120 @@ class Blocker {
...
+
+    // we don't want to make any BlockerT to be aware and coupled with
+    // an operation. to not templatize an entire path from an op to
+    // a blocker, type erasuring is used.
+    struct TriggerI {
+      template <class FutureT>
+      void maybe_record_blocking(const FutureT& fut, const T& blocker) {
+        if (!fut.available()) {
+          // a full blown call via vtable. that's the cost for templatization
+	  // avoidance. anyway, most of the things actually have the type
```
Comment (rzarzynski) [inline src/crimson/common/operation.h:234]: Fix the indentation (here and in other places as well).

File: src/crimson/common/operation.h [hunk 2]
```diff
@@ -411,6 +464,13 @@ class PipelineStageI : public Blocker {
   virtual seastar::future<PipelineExitBarrierI::Ref> enter() = 0;
 };
 
+
+template <class T>
+class PipelineStageIT : public BlockerT<T> {
```
Comment (rzarzynski) [inline src/crimson/common/operation.h:469]: Drop the non-templated `PipelineStageI`.

**Comment by athanatos** [comment]
Misc build fixes: https://github.com/athanatos/ceph/commit/f3752af0f925e4476c6e9077a22e6a8165bd7cc4

clang really wants the template keyword for dependent types and there are some missing virtual destructors.  There's also a workaround for a linker error for LTTNGBackend.  Can we merge part of this relatively soon?  I've got some conflicting changes I'm rebasing on top of it.

**Comment by rzarzynski** [comment]
Thanks for the patch, @athanatos! The recording part is basically complete (a couple of small `TODOs` I'm addressing today). The `dump_ops_in_flight` is also there, so I think not much core stuff left and we should be able to roll out this week.

`LttngBackend` can be postponed to a follow-up PR.

File: src/crimson/common/operation.h [hunk 3]
```diff
@@ -155,16 +158,188 @@ class Blocker {
...
+      }
+
+      utime_t timestamp;
+      const T* blocker;
+    } internal_backend;
+
+    // we don't want to make any BlockerT to be aware and coupled with
+    // an operation. to not templatize an entire path from an op to
+    // a blocker, type erasuring is used.
+    struct TriggerI {
```
Comment (athanatos) [inline src/crimson/common/operation.h:251]: Needs a virtual destructor.

**Comment by rzarzynski** [comment]
```
99% tests passed, 1 tests failed out of 277

Total Test time (real) = 1412.37 sec

The following tests FAILED:
	239 - unittest-transaction-manager (Failed)
```

**Comment by rzarzynski** [comment]
jenkins retest this please

**Comment by rzarzynski** [comment]
Fixed a life-time management in exit barrier tracking, dropped the DNM / WIP commits.

**Comment by rzarzynski** [comment]
Fixed another LT. `unittest-transaction-manager` is back green.

**Comment by rzarzynski** [comment]
Linking the Sam's pipeline refactor for multi-reactor: #46136.

**Review by athanatos — APPROVED** [review]

**Pull request merged**